dial/pipeline.go

```diff
@@ -0,0 +1,313 @@
+	return &Request{
+		ctx:      ctx,
+		net:      net,
+		id:       id,
+		notifyCh: make(chan struct{}),
+		status:   requestInflight,
+	}
+}
+
+// Complete fills in the result of the dial request. It can only be invoked once per Request.
+// After the first completion, further calls to complete will fail with error.
+//
+// Complete invokes all callbacks associated with this Request, and closes the notifyCh.
+func (req *Request) Complete(conn inet.Conn, err error) error {
+	if !atomic.CompareAndSwapInt32(&req.status, requestInflight, requestFinished) {
+		return ErrDialRequestInvalidStatus
+	}
+
+	req.conn, req.err = conn, err
+	for i := len(req.callbacks); i > 0; i-- {
+		req.callbacks[i-1]()
+	}
+
+	close(req.notifyCh)
+	return nil
+}
+
+func (req *Request) IsComplete() bool {
+	return atomic.LoadInt32(&req.status) == requestFinished
+}
+
+// Values returns the connection and error fields from this request.
+// Both return values may be nil or incoherent unless the request has completed (see IsComplete).
+func (req *Request) Values() (inet.Conn, error) {
+	return req.conn, req.err
+}
+
+// AddCallback adds a function that will be invoked when this request completes, either in success
+// or in failure.
+func (req *Request) AddCallback(cb func()) {
+	req.callbacks = append(req.callbacks, cb)
+}
+
+// Job represents a dial attempt to a single multiaddr. It is associated 1-* to a Request.
+// It can carry its own context and its own callbacks.
+type Job struct {
+	status int32
+	req    *Request
+
+	ctx  context.Context
+	addr ma.Multiaddr
+
+	callbacks []func()
+
+	// result.
+	tconn tpt.Conn
+	err   error
+
+	// Channel where completed dials should be sent.
+	completeCh chan *Job
+}
+
+func NewDialJob(ctx context.Context, req *Request, addr ma.Multiaddr) *Job {
+	return &Job{req: req, ctx: ctx, addr: addr, status: jobInflight}
+}
+
+func (j *Job) Cancelled() bool {
+	select {
+	case <-j.req.ctx.Done():
+		return true
+	case <-j.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (j *Job) AddCallback(cb func()) {
+	j.callbacks = append(j.callbacks, cb)
+}
+
+func (j *Job) Complete(conn tpt.Conn, err error) error {
+	if !atomic.CompareAndSwapInt32(&j.status, jobInflight, jobFinished) {
+		return ErrDialJobInvalidStatus
+	}
+
+	j.tconn, j.err = conn, err
+	for i := len(j.callbacks); i > 0; i-- {
+		j.callbacks[i-1]()
+	}
+
+	j.completeCh <- j
+	return nil
+}
+
+type dialJobs []*Job
+
+func (djs *dialJobs) sift() (success dialJobs, failed dialJobs) {
+	for _, dj := range *djs {
+		if dj.err == nil {
+			success = append(success, dj)
+			continue
+		}
+		failed = append(failed, dj)
+	}
+	return success, failed
+}
+
+type AddConnFn func(tc tpt.Conn, dir inet.Direction) (inet.Conn, error)
+
+type Pipeline struct {
+	lk  sync.RWMutex
+	ctx context.Context
+
+	net inet.Network
+
+	reqPreparers []RequestPreparer
+	planner      Planner
+	throttler    Throttler
```

Comment:
Personally, I'd allow multiple of these. Throttler's can easily throttle a single job type, passing the rest on to the next throttler.

Reply:
Hmmm... would there be a case to have multiple `Planner`s as well? Rather than modelling the cardinality here, we could provide `Composite*` wrappers that from the outside behave the same, with their constructors accepting a number of Planners, Throttlers, etc.

We could even kill the n-cardinality on the `{Request/Job}Preparers`, in favour of composites. WDYT?